### ISSUE TITLE
Use document provider to grant third-party apps access to app-private directories

### ISSUE THREAD

**Issue by MuntashirAkon** [MuntashirAkon/AppManager#516]
With the introduction of File Manager where an elegant method has been developed to support opening any files from any directory via a content provider so that the recipient app does not require any file or root permission, it is possible to enhance this support via a document provider so that a third-party app can access app-specific private directories such as `/data/data/package.name` without root permission. This could greatly reduce the attack surface as fewer number of apps would need to use the root permission.

Possible use cases:

1. Editing XML files at `/data/system/ifw/` without granting root or even the editors that do not support root but support SAF (AKA document provider).
2. Granting file managers access to certain app directory, such as `/data/data/io.github.muntashirakon.AppManager/files/profiles` so that you can add/edit profiles directly.
3. Backing up certain directory of interest via third-party event handling apps such as Tasker without letting them know about the location of the files and/or without granting root
4. I forgot others, but the general idea is to minimise apps that can access root.

**Comment by d4rkk3y** [comment]
> With the introduction of File Manager where an elegant method has been developed to support opening any files from any directory via a content provider so that the recipient app does not require any file or root permission

I'm a bit confused about "does not require file or root permission". I mean don't need root device or don't need grant root permission but still need root device.

**Comment by MuntashirAkon** [comment]
> I'm a bit confused about "does not require file or root permission". Does it mean don't need root device or don't need grant root permission but still need root device?

AM, of course, would require root permission. How else is AM going to access files and folders accessible only via root? The purpose is to minimise root usage by third-party apps in order to minimise attack surface, allowing the AM users handle app-private data with their favourite file manager or editor app without the need to grant root for the file manager. Document provider is perfect for this because it is an abstract file accessing mechanism (the core part of the Storage Access Framework) where the target app do not know the exact location of the file being processed.

**Comment by d4rkk3y** [comment]
> AM, of course, would require root permission. How else is AM going to access files and folders accessible only via root? The purpose is to minimise root usage by third-party apps in order to minimise attack surface, allowing the AM users handle app-private data with their favourite file manager or editor app without the need to grant root for the file manager. Document provider is perfect for this because it is an abstract file accessing mechanism (the core part of the Storage Access Framework) where the target app do not know the exact location of the file being processed.

I see. Btw, maybe it will not very useful for most users but with adb mode users still can access "debuggable" third-party app-private data right? Can we use app_process run a service with that debuggable app permission to access its app-private data?

**Comment by MuntashirAkon** [comment]
> maybe it will not very useful for most users

Well, it mostly depends on my usage than the users. If I see something useful for me, I implement it in order to suite my need. Besides, many users are incapable of understanding the global picture of a feature unless it is implemented.

> with adb mode users still can access "debuggable" third-party app-private data right?

Correct. The `run-as` command works for both root and ADB users.

> Can we use app_process run a service with that debuggable app permission to access its app-private data?

I don't know what you've exactly meant here. But `app_process` doesn't work in the context of a user app, but if you take a look at the [`run-as` source code](https://cs.android.com/android/platform/superproject/+/master:system/core/run-as/run-as.cpp), you'd see that it involves changing SELinux context to the target user app, effectively the same as `su - <uid>` in Magisk's `su` binary except that the former changes the working directory to the app-private directory before starting a shell environment. One solution could be to get this shell process in our remote service and use [libsu's IO library](https://github.com/topjohnwu/libsu) for communication but the trouble is `shell` is notoriously slow, and for listing files we'd have to use the unreliable `ls` command.

**Comment by d4rkk3y** [comment]
> One solution could be to get this shell process in our remote service and use libsu's IO library for communication but the trouble is shell is notoriously slow, and for listing files we'd have to use the unreliable ls command.

Is it possible to run a server/daemon(which implement file manager code)  in this shell process with `app_process` and use IPC to access it function? I am sorry if it is stupid question because of my limited knowledge about this.

**Comment by MuntashirAkon** [comment]
> Is it possible to run a server/daemon(which implement file manager code) in this shell process with `app_process` and use IPC to access it function?

No. As I've said above, `app_process` doesn't work for user apps.

**Comment by JustAnUserLol** [comment]
It is not possible to access `/data/com.pacakage.name` with document provider, only accessing `/storage/emulated/0/com.package.name` is possible.

**Comment by null-dev** [comment]
For **non-rooted** users stumbling over this, note that Revanced is able to inject this feature into arbitrary applications: https://github.com/ReVanced/revanced-patches/pull/3830

It isn't working at the moment due to [another bug](https://github.com/ReVanced/revanced-patches/issues/4058), but I'll try to look into it when I have some time.

Also, about this:

> > Is it possible to run a server/daemon(which implement file manager code) in this shell process with `app_process` and use IPC to access it function?
> 
> No. As I've said above, `app_process` doesn't work for user apps.

I stumbled across this very interesting project: https://github.com/square/stoic

It injects into a debuggable application and can:
- Injects arbitrary code
- Injects a debug server
- Setup a bi-directional multiplexed data stream over `adb shell`

BI-DI over `adb shell` may still be slow but since you can inject arbitrary code, you can probably also just start a TCP server and then do any further communication over that.

And also, apparently you can inject [Frida](https://frida.re/) this way too: https://koz.io/library-injection-for-debuggable-android-apps/

Anyways, this is a very niche feature so not expecting you to work on this Muntashir, but I thought it was really cool and kinda relevant so I wanted to document it in case someone else finds it useful.

**Comment by Tharow** [comment]
@MuntashirAkon Something like this? from AOSP File Browser
![Image](https://github.com/user-attachments/assets/11b9fa8c-b882-4f7c-9102-970adbab49a2)

**Comment by MuntashirAkon** [comment]
> Something like this? from with-in AOSP File Browser

Not yet sure.

**Comment by furrydaemon** [comment]
Hi, I dunno if I could add
I use the file functionality of App Manager in ADB mode and I can look into the Android/data/ directory
However, if I try to read a picture or a video using a third party app it wont read it

I can use alternatively the document provider to access those folders and i can read the files, i dunno if a fix is planned, i hope so
